[WPE][GTK] web process leak if webkit_download_set_destination is called with empty destination

yury-s opened the pull request:
#### c985acc37157f1ac8e7601e494ebd71bd76e6509
<pre>
[WPE][GTK] web process leak if webkit_download_set_destination is called with empty destination
<a href="https://bugs.webkit.org/show_bug.cgi?id=242217">https://bugs.webkit.org/show_bug.cgi?id=242217</a>

Reviewed by Michael Catanzaro.

Fail if download destination is set to &quot;&quot;.

 * Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp:
  (webkit_download_set_destination):

Canonical link: <a href="https://commits.webkit.org/256737@main">https://commits.webkit.org/256737@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a37d8abfd8f7bb2e62312879b5ec76ba1073bf4e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96608 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5864 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29685 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106134 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6060 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34602 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88994 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/102845 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102283 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/4518 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83211 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31499 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/86361 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88241 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/74414 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/40333 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19732 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37996 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21136 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4677 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/4265 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/43676 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1078 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40412 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

yury-s commented on Source/WebKit/NetworkProcess/soup/NetworkDataTaskSoup.cpp:
```diff
@@ -521,7 +521,11 @@ void NetworkDataTaskSoup::dispatchDidReceiveResponse()
-            clearRequest();
+            if (isDownload()) {
+                invalidateAndCancel();
+                dispatchDidCompleteWithError(downloadCancelledByUserError(m_response));
```
Looking at what Cocoa port does I believe it may make sense to run this code unconditionally when the policy action is ignore.

mcatanzaro replied:
I know it's been a while, but any chance you remember why you suggested this? `NetworkDataTaskCocoa::didReceiveResponse` doesn't look anything like this.

yury-s replied:
Unfortunately I don't remember already. In any case I reverted this part of the change as you suggested and replaced with g_return_if_fail in the api implementation.

mcatanzaro commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestDownloads.cpp:
```diff
@@ -827,6 +827,31 @@ static void testDownloadEphemeralContext(Test* test, gconstpointer)
+    GRefPtr<WebKitDownload> download = adoptGRef(webkit_web_context_download_uri(context.get(), kServer->getURIForPath("/test.pdf").data()));
+    test->assertObjectIsDeletedWhenTestFinishes(G_OBJECT(download.get()));
+    g_signal_connect(download.get(), "decide-destination", G_CALLBACK(+[](WebKitDownload* download, const gchar* suggestedFilename, gpointer) {
+        webkit_download_set_destination(download, "");
```
Eh, is this really expected to work? The API does not say that passing an empty string is allowed. I would expect a critical.

Also: you need to return TRUE here, otherwise it's undefined behavior.

yury-s replied:
> Eh, is this really expected to work? The API does not say that passing an empty string is allowed. I would expect a critical.

Dunno but the implementation does [handle that](https://github.com/WebKit/WebKit/blob/fcd1d36de2ca84f2d8736165fbc08cd6251f9578/Source/WebKit/NetworkProcess/NetworkProcess.cpp#L2042-L2043) explicitly.

> Also: you need to return TRUE here, otherwise it's undefined behavior.

Fixed. I copied the code from the test above which didn't have such statement.

mcatanzaro commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestDownloads.cpp:
```diff
@@ -816,6 +816,7 @@ static void testDownloadEphemeralContext(Test* test, gconstpointer)
         GUniquePtr<char> destination(g_build_filename(Test::dataDirectory(), suggestedFilename, nullptr));
         GUniquePtr<char> destinationURI(g_filename_to_uri(destination.get(), nullptr, nullptr));
         webkit_download_set_destination(download, destinationURI.get());
+        return true;
```
Careful, this needs to be `return TRUE` since it's a `gboolean` (`int`, not a `bool`)

mcatanzaro commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestDownloads.cpp:
```diff
@@ -827,6 +828,32 @@ static void testDownloadEphemeralContext(Test* test, gconstpointer)
+    test->assertObjectIsDeletedWhenTestFinishes(G_OBJECT(download.get()));
+    g_signal_connect(download.get(), "decide-destination", G_CALLBACK(+[](WebKitDownload* download, const gchar* suggestedFilename, gpointer) {
+        webkit_download_set_destination(download, "");
+        return true;
```
Here too.

mcatanzaro commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestDownloads.cpp:
```diff
@@ -827,6 +828,32 @@ static void testDownloadEphemeralContext(Test* test, gconstpointer)
+        webkit_download_set_destination(download, "");
+        return true;
+    }), nullptr);
+    g_signal_connect(download.get(), "finished", G_CALLBACK(+[](WebKitDownload*, GMainLoop* loop) {
```
Code style suggestion:

```suggestion
    }), nullptr);

    g_signal_connect(download.get(), "finished", G_CALLBACK(+[](WebKitDownload*, GMainLoop* loop) {
```

mcatanzaro reviewed (commented):
So I have an alternative suggestion: let's just add a `g_return_if_fail(uri[0] != '\0')` to the top of `webkit_download_set_destination()` and move on? If the application wishes to cancel the download, it could just use `webkit_download_cancel()`: no need to call `webkit_download_set_destination()` with an empty string. That's a lot more explicit.

yury-s committed:
[WPE][GTK] web process leak if webkit_download_set_destination is called with empty destination
https://bugs.webkit.org/show_bug.cgi?id=242217

Reviewed by Michael Catanzaro.

Fail if download destination is set to "".

 * Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp:
  (webkit_download_set_destination):

Canonical link: https://commits.webkit.org/256737@main
